### DISCUSSION TITLE
How to detect unobserved fingerprint bits in a molecule, based on a reference set

### DISCUSSION THREAD

**Discussion by GLPG-GT** [rdkit/rdkit#6509] [Q&A]
Suppose you have a (large) set of reference smiles, and a set of test smiles.  
You want to know if any Morgan radius 1 fingerprint bits exist in the test molecules, which were not observed in the reference molecules.

Here is what I could put together so far:

```
from rdkit import Chem
from rdkit.Chem import rdFingerprintGenerator
from rdkit import DataStructs
import rdkit
```
```
mfpgen = rdFingerprintGenerator.GetMorganGenerator(radius = 1)
ao = rdFingerprintGenerator.AdditionalOutput()
ao.AllocateBitInfoMap()
```
```
reference_smiles = ['CCC', 'CCN', 'CCO']
reference_dict = dict()
for sm in reference_smiles :
    m = Chem.MolFromSmiles(sm)
    sfp = mfpgen.GetSparseFingerprint(m, additionalOutput = ao)
    print(sfp.GetNumOnBits())
    print(list(sfp.GetOnBits()))
    bitinfomap = ao.GetBitInfoMap()
    bits = list(bitinfomap.keys())
    print(bits)
    for b in bits :
        if b not in reference_dict :
            reference_dict[b] = 1
        else :
            reference_dict[b] += 1
```
4
[-2049583024, -2048238559, -752510682, 2068133184]
[2068133184, 2245384272, 2246728737, 3542456614]
6
[-2049583024, -2048238559, -1702181931, -752510682, 772817685, 847957139]
[772817685, 847957139, 2245384272, 2246728737, 2592785365, 3542456614]
6
[-2049583024, -2048238559, -752510682, -276918910, 864662311, 1535166686]
[864662311, 1535166686, 2245384272, 2246728737, 3542456614, 4018048386]

[Side note: why are some of the bits generated by `GetOnBits()` different from those listed by `GetBitInfoMap()`? In particular the ones encoded by a negative integer. But OK, that's not the main question].

```
test_smiles = ['CCS', 'OCCS']
for sm in test_smiles :
    print('\nTest smiles = ' + sm)
    m = Chem.MolFromSmiles(sm)
    sfp = mfpgen.GetSparseFingerprint(m, additionalOutput = ao)    
    print('> # bits = ' + str(sfp.GetNumOnBits()))
    #print(list(sfp.GetOnBits()))
    bitinfomap = ao.GetBitInfoMap()
    bits = list(bitinfomap.keys())
    print('bits = ' + str(bits))
    unknown_bits = []
    for b in bits :
        if b not in reference_dict :
            unknown_bits.append(b)
    nub = len(unknown_bits)
    if nub > 0 :
        usms = []
        for ub in unknown_bits :
            tu = bitinfomap[ub]
            tu0 = tu[0]
            a = tu0[0]
            r = tu0[1]
            env = Chem.FindAtomEnvironmentOfRadiusN(m, r, a)
            atoms = set()
            if r > 0 :
                for bidx in env :
                    atoms.add(m.GetBondWithIdx(bidx).GetBeginAtomIdx())
                    atoms.add(m.GetBondWithIdx(bidx).GetEndAtomIdx())
                usm = Chem.MolFragmentToSmiles(m, atomsToUse = list(atoms), bondsToUse = env, rootedAtAtom = a)
            else :
                atoms.add(a)
                usm = Chem.MolFragmentToSmiles(m, atomsToUse = list(atoms), bondsToUse = [], rootedAtAtom = a)
            usms.append(usm)
        print('> # unknown bits = ' + str(nub))
        print('> unknown bits = ' + str(unknown_bits))
        print('> unknown subs = ' + str(usms))
```
Test smiles = CCS
\> \# bits = 6
bits = [751130035, 1026924773, 2245384272, 2246728737, 3542456614, 4096536289]
\> \# unknown bits = 3
\> unknown bits = [751130035, 1026924773, 4096536289]
\> unknown subs = ['SC', 'S', 'C(C)S']

Test smiles = OCCS
\> \# bits = 7
bits = [751130035, 864662311, 1026924773, 1535166686, 2245384272, 4023654873, 4118913676]
\> \# unknown bits = 4
\> unknown bits = [751130035, 1026924773, 4023654873, 4118913676]
\> unknown subs = ['SC', 'S', 'C(C)O', 'C(C)S']

Ideally, we would make a reference dict from a very large set of smiles, store it as a simple text file, and remake the dictionary from it to use it on several test molecules, when necessary (without having to reprocess the reference smiles each time).

> **Do you see anything wrong with the above? Is a dictionary of bits the most efficient way to do this, or would you suggest anything else?**

It would also be interesting to _show_ the unobserved bits in the test molecule.  
Still not sure what the best way would be. This is what I tried:

```
from rdkit.Chem import Draw
mfp2_svg = Draw.DrawMorganBit(m, 4118913676, bitinfomap, useSVG = True)
mfp2_svg
```
![download](https://github.com/rdkit/rdkit/assets/96192927/46c56281-f4f6-441f-a96b-84e4d7f69081)

However this is only one bit; what if we wanted a svg with all the unknown bits?  
Or alternatively, the original molecule with the atoms and bonds corresponding to the unknown bits highlighted?

**Comment by greglandrum** [comment]
@GLPG-GT the easiest way to figure out which bits are set in one fingerprint but not another is to use set operations on the results from the `fp.GetOnBits()` function. Here's a simple demo:
```
In [12]: m1 = Chem.MolFromSmiles('Cc1cccnc1')

In [13]: m2 = Chem.MolFromSmiles('c1cccnc1')

In [14]: fpg = rdFingerprintGenerator.GetMorganGenerator(radius=2)

In [15]: fp1 = fpg.GetSparseFingerprint(m1)

In [16]: fp2 = fpg.GetSparseFingerprint(m2)

# bits set in fp1 and not in fp2:
In [17]: set(fp1.GetOnBits()).difference(fp2.GetOnBits())
Out[17]: 
{-2048238559,
 -1631890855,
 -1087400161,
 -1077586588,
 -517798401,
 -50791393,
 422715066,
 951226070,
 1153219603,
 1194681160,
 1349404210}

# bits set in fp2 but not fp1:
In [18]: set(fp2.GetOnBits()).difference(fp1.GetOnBits())
Out[18]: {-1531113083, 1207774339, 1343371647}
```

You can keep track of all the bits seen in a group of molecules by using the set's `update()` function:
```
In [19]: all_bits = set(fp1.GetOnBits())

In [20]: all_bits.update(fp2.GetOnBits())

In [21]: set(fp1.GetOnBits()).difference(all_bits)
Out[21]: set()
```

**Reply by GLPG-GT** [reply]
Thanks @greglandrum !
If at some point we decide that we only want to detect the missing bits (as in their numerical identifiers), we will indeed use `GetOnBits()` with `set()`; I suppose it's faster than looking up a dictionary.
Then practically we will still need to store the set bits in a file (or something to that effect), as we won't be reprocessing the reference set each time (it's close to 2 million molecules, it takes quite some time).

However, we do also need to know which **substructures**/features of each test molecule are not found in the reference.
And, as far as I can glean from the documentation and examples, that can only be done using the bits returned by `GetBitInfoMap()`.
But if I am wrong, and the `GetOnBits()` _can_ be used to generate the smiles of the unobserved substructures, I am certainly keen to try that out.
